src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1990,48 +2019,47 @@ Symbol* SystemDictionary::check_signature_loaders(Symbol* signature,
 Method* SystemDictionary::find_method_handle_intrinsic(vmIntrinsicID iid,
                                                        Symbol* signature,
                                                        TRAPS) {
+
+  MutexLocker ml(THREAD, InvokeMethodTable_lock);
```

Comment:
Since this is a refactoring RFE, I think we should avoid substantial logic changes, especially ones related to locking. 

If we hold the `InvokeMethodTable_lock` lock the whole time, it will prevent two concurrent threads from calling `Method::make_method_handle_intrinsic` on two different `iid`. This seems to be an optimization that the original code tries to do (or unintentionally allows), with the trade off being that if two threads try to use the same ``iid`, we might make two method handles and throw away one of them. The unused one might be garbage collected though. Without looking at the details, I can't confirm that a leak would happen, or if it does, how often.

In any case, if we think that such a change is necessary, it should be deferred to a follow-up PR.

If I understand correctly, the original code can be done without a lock, because the old Hashtable allows lock-free reads, but ResourceHashtable doesn't.

```
SymbolPropertyEntry* spe = invoke_method_table()->find_entry(index, hash, signature, iid_as_int);
```

I would suggest changing the new code to do the same two-step locking as done in the new version of `SystemDictionary::find_method_handle_type()`, but you might need to hold `SystemDictionary_lock` instead. Please see my comments for `SystemDictionary::methods_do()`.